// product-scraper/index.js
var Xray = require('x-ray')
	x = Xray(),
	xDelay = Xray().delay('1s', '10s'),
	URL = require('url-parse'),
	cheerio = require('cheerio')
	request = require('request'),
	probe = require('probe-image-size');

exports.images = function (url, callback, scope) {

	request(url, function (err, resp, html) {

		var $ = cheerio.load(html);
		var minThreshold = 400;
		var scope = (scope) ? scope : 'body';

		var _images = [], _index = 0, totalImages = $(scope).find('img[src]').length;

		$(scope).find('img[src]').each(function () {
			var src = $(this).attr('src');
			probe({ url: src, timeout: 2500 }, function (err, result) {
				if (result && result.mime !== 'image/gif' && result.width > minThreshold) {
					_images.push({
						width: result.width,
						height: result.height,
						src: src
					});
				}

				_index++;
				if (_index == totalImages) {
					_images.sort(function (a, b) {
						return b.width - a.width;
					});

					_images = _images.splice(0, 4);

					if (typeof callback == 'function') callback(_images);
				}
			});
		});

	});

}

exports.sites = {
	base: function () {
		return {
			brand: '',
			description: '',
			details: '',
			ean: '',
			id: '',
			image: '',
			isbn: '',
			model: '',
			original_price: '', // Ancien prix si promotion
			price: '', // Prix si pas de promotion
			sale_price: '',  // Nouveau prix si promotion
			tags: '', 
			title: ''
		};
	},
	amazon: function (product_id) {

		var productSelector = this.base();

		productSelector.id = '#ASIN@value';
		productSelector.isbn = '#ASIN@value';
		productSelector.title = '#productTitle';
		productSelector.original_price = x('.a-text-strike');
		productSelector.image = xDelay('#landingImage@data-old-hires');
		productSelector.price = '#priceblock_ourprice';
		productSelector.brand = '#brand';
		productSelector.details = x(['#feature-bullets > ul > li']);
		productSelector.description = '#productDescription';

		var pageURL = 'http://www.amazon.com/' + product_id;

		return {
			page: pageURL,
			selectors: productSelector,
			images: true
		};

	},
	asos: function (path) {

		var productSelector = this.base();

		productSelector.id = 'div.product-code > span';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '';
		productSelector.price = 'p.total-price'; // Marche pas, tout est bind en JS
		productSelector.sale_price = '';
		productSelector.brand = '';
		productSelector.details = x(['']);
		productSelector.description = ''; 
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.asos.fr/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	bestbuy: function (path) {

		var productSelector = this.base();

		productSelector.id = '#sku-value';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '.regular-price';
		productSelector.price = '.item-price';
		productSelector.brand = '#schemaorg-brand-name@content';
		productSelector.details = x(['.item-price']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');
		productSelector.tags = 'meta[name="keywords"]@content';

		var pageURL = 'http://www.bestbuy.com/site/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	castorama: function (path) {

		var productSelector = this.base();

		productSelector.id = '.refNum';
		productSelector.title = '.productContent > h1';
		productSelector.original_price = '.oldprice';
		productSelector.price = '.price'; 
		productSelector.sale_price = '.newprice';
		productSelector.brand = '.productLabel > a > img@title';
		productSelector.details = x(['.tabTechnique > ul > li']);
		productSelector.description = 'meta[name="description"]@content';
		productSelector.image = xDelay('.productImage > a > img@src');
		productSelector.tags = 'meta[name="keywords"]@content';

		var pageURL = 'http://www.castorama.fr/store/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	cdiscount: function (path) { // MARCHE PAS

		var productSelector = this.base();

		productSelector.id = '';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '';
		productSelector.price = '';
		productSelector.brand = '';
		productSelector.details = x(['']);
		productSelector.description = 'meta[name="description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');
		productSelector.tags = 'meta[name="keywords"]@content';

		var pageURL = 'http://www.cdiscount.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	darty: function (path) {

		var productSelector = this.base();

		productSelector.id = '#main > div@data-codic'
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = 'div.product_column.product_right > div.product_infos.clearfix > div.product_metas > div.product_price_infos > div > span';
		productSelector.price = 'div.product_column.product_right > div.product_infos.clearfix > div.product_metas > div.product_price.font-novecentosSWBD > span';
		productSelector.sale_price = 'div.product_column.product_right > div.product_infos.clearfix > div.product_metas > div.product_price.font-novecentosSWBD > span';
		productSelector.details = x(['div.product_column.product_right > div.product_details > ul']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.darty.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	fnac: function (path) {

		var productSelector = this.base();

		productSelector.id = 'body@data-prid';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '.oldPrice';
		productSelector.price = '.product-price';
		productSelector.brand = 'span[itemprop="manufacturer"]';
		productSelector.details = x(['.whiteContent > ul > li']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.fnac.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	grosbill: function (path) {

		var productSelector = this.base();

		productSelector.id = '#PopinDiv@data-product';
		productSelector.title = 'h1[itemprop="name"]';
		productSelector.original_price = 'div.datasheet_price_and_strike_price_wrapper > div:nth-child(2) > b > s';
		productSelector.price = 'meta[itemprop="price"]'; // Même valeur que sale_price
		productSelector.sale_price = 'div.datasheet_price_and_strike_price_wrapper > div:nth-child(1) > b'; // Même valeur que price
		productSelector.brand = '#datasheet_wrapper > div:nth-child(12) > div > table > tbody > tr:nth-child(4) > td:nth-child(2)'; // Voir pour faire mieux
		productSelector.details = x(['#datasheet_wrapper > div.gb_container > div > div:nth-child(2) > p']); // Voir pour faire mieux
		productSelector.description = '.datasheet_info_wrapper'; 
		productSelector.image = xDelay('#zoom_datasheet_product[itemprop="image"]@src');

		var pageURL = 'https://www.grosbill.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	laredoute: function (path) {

		var productSelector = this.base();

		productSelector.id = '#pdpFRdivMain@data-productid';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.sale_price = '.final-price';
		productSelector.original_price = '.sale-price-before';
		productSelector.price = '.final-price';
		productSelector.brand = '#ctl00_cphMain_ctl00_ctl00_aBrand';
		productSelector.details = x(['.left > div']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.laredoute.fr/ppdp/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	leroymerlin: function (path) {

		// Doesnt Work, le html n'est pas chargé ou quel que chose comme ça
		// 	var productSelector = this.base();
		// 	/* Toute les infos se trouve dans l'objet tc_vars en JS */
		// 	productSelector.id = '#global-reflm > span';
		// 	// ou productSelector.id = 'tc_vars.product_id'; 
		// 	productSelector.title = 'meta[property="og:title"]@content';
		// 	productSelector.original_price = '#corps > div.centerContent > article > aside > div.price-wrapper.decli-prix-principal.promo > p.infos > em > em';
		// 	productSelector.price = '#corps > div.centerContent > article > aside > div.price-wrapper.decli-prix-principal.promo > p.price > strong';
		// 	productSelector.details = x(['#description-detaillee']);
		// 	productSelector.description = 'meta[property="og:description"]@content';
		// 	productSelector.image = xDelay('meta[property="og:image"]@content');
		// 	console.log(productSelector);

		var productSelector = this.base();

		productSelector.id = ''; 
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '';
		productSelector.price = '';
		productSelector.brand = '';
		productSelector.details = x([]);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.leroymerlin.fr/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	livrefnac: function (path) {

		var productSelector = this.base();

		productSelector.id = 'body@data-prid';
		productSelector.isbn = 'meta[property="og:isbn"]@content';
		productSelector.title = 'meta[property="og:title"]@content';
		productSelector.original_price = '.oldPrice';
		productSelector.price = '.product-price';
		productSelector.details = x(['.whiteContent > ul > li']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://livre.fnac.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};
	},
	ldlc: function (path) {

		var productSelector = this.base();

		productSelector.id = ''; //PB00210019, j'arrive pas à le prendre il est dans l'url et dans des liens dans le DOM
		productSelector.title = 'span[itemprop="name"]';
		productSelector.original_price = 'span.prix > span.refPrice';
		productSelector.price = 'meta[itemprop="og:price"]@content';
		productSelector.sale_price = 'span.prix > span.price.sale';
		productSelector.brand = 'meta[itemprop="name"]@content';
		productSelector.model = 'span[itemprop="mpn"]@content',
		productSelector.details = x(['span.[itemprop="description"]']);
		productSelector.description = 'span.designation_longue'; 
		productSelector.image = xDelay('meta[name="og:image"]@content');

		var pageURL = 'http://www.ldlc.com/fiche/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	rueducommerce: function(path) { 

		var productSelector = this.base();

		productSelector.id = '#BVRRContainer > div > div@data-product-id'; 
		productSelector.isbn = 'meta[property="og:isbn"]@content';
		productSelector.title = 'h1';
		productSelector.original_price = '.retail > p.price';
		productSelector.sale_price = 'meta[itemprop="price"]@content';
		productSelector.price = 'meta[itemprop="price"]@content';
		productSelector.details = x(['#blocAttributesContent']);
		productSelector.description = 'meta[property="og:description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');

		var pageURL = 'http://www.rueducommerce.fr/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	target: function (path) {

		var productSelector = this.base();

		productSelector.title = x('#ProductDetailsTop > div:nth-child(6) > div.contentRight.primaryImgContainer > h2 > span');
		productSelector.image = xDelay('#heroImage@src');
		productSelector.price = '#price_main > div > span > span';
		productSelector.description = '#item-overview > div > div:nth-child(1) > div:nth-child(1) > div > p > span';
		productSelector.details = x(['#item-overview > div > div:nth-child(1) > div:nth-child(1) > ul li']);

		var pageURL = 'http://www.target.com/p/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	tesco: function (path) {

		var productSelector = this.base();

		productSelector.id = 'meta[property="og:upc"]@content';
		productSelector.title = '#productLabel';
		productSelector.price = '.current-price[itemprop="price"]'
		productSelector.details = x(['.main-details > ul > li']);
		productSelector.description = 'meta[name="description"]@content';
		productSelector.image = xDelay('.s7staticimage > img@src');

		var pageURL = 'http://www.tesco.com/direct/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},
	ubaldi: function(path) { 

		var productSelector = this.base();

		productSelector.ean = '#dv_ean@title';
		productSelector.model = '#dv_ref@title';
		productSelector.id = '#wc-power-page@data-product-id'; 
		productSelector.title = 'h1 .fa-titre-article';
		productSelector.original_price = '.prix-lancement@data-prix-lancement';
		productSelector.sale_price = '.fa-prix > span';
		productSelector.price = '.fa-prix > span';
		productSelector.details = x(['.liste-specs > div']);
		productSelector.description = 'meta[name="description"]@content';
		productSelector.image = xDelay('meta[property="og:image"]@content');
		productSelector.tags = 'meta[name="keywords"]@content';

		var pageURL = 'http://www.ubaldi.com/' + path;

		return {
			page: pageURL,
			selectors: productSelector
		};

	},

};

exports.scraper = function (opts, callback) {

	var data, lookup, self = this;

	if (this.sites[opts.site]) {
		lookup = this.sites[opts.site](opts.product_id);

		x(lookup.page, lookup.selectors)
			(function (err, obj) {
				var _obj = obj;
				_obj.url = lookup.page;

				if (_obj.details) {
					var details = [];
					_obj.details.forEach(function (element, index, array) {
						details.push(element
							.replace(/(?:\t|\r|\n)/g, '')
							.replace(/ +(?= )/g, ''));
					});
					_obj.details = details;
				}

			if(_obj.id) _obj.id = _obj.id.trim();
			// RAJOUTER CONDITION POUR CASTORAMA 
			if(_obj.id) _obj.id = _obj.id.trim()
					.replace('Réf :', '');
			if(_obj.isbn) _obj.isbn = _obj.isbn.trim(); 
			if(_obj.original_price) _obj.original_price = _obj.original_price.trim(); 
			// RAJOUTER CONDITION POUR BESTBUY
			if(_obj.original_price) _obj.original_price = _obj.original_price.trim().replace('(Reg. ', '').replace(')', ''); 
			// RAJOUTER CONDITION POUR DESCRIPTIONS AMAZON
			if(_obj.description) _obj.description = _obj.description.trim()
					.replace(/(?:\t|\r|\n)/g, '')
					.replace(/ +(?= )/g,'');
			if(_obj.title) _obj.title = _obj.title.trim();
			if(_obj.price) _obj.price = _obj.price.trim();
			if(_obj.brand) _obj.brand = _obj.brand.trim();

				if (lookup.images) {
					self.images(_obj.url, function (images) {
						_obj.images = images;
						callback(_obj);
					});
				}
				else {
					callback(_obj);
				}

			});
	}
	else {
		var selectors = {
			id: '[itemprop="product-id"]',
			brand: '[itemprop="brand"]',
			description: '[itemprop="description"]',
			title: 'meta[property="og:title"]@content',
			image: xDelay('[itemprop="image"]@src'),
			price: '[itemprop="price"]'
		};

		x(opts.url, selectors)
			(function (err, obj) {

				self.images(opts.url, function (images) {
					obj.images = images;
					callback(obj);
				});

			});
	}
}

exports.parseURL = function (url, callback) {

	var parse = new URL(url);
	
	if (parse.host == 'www.amazon.com') {
		this.scraper({
			site: 'amazon',
			product_id: parse.pathname.replace('/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.asos.fr') {
		this.scraper({
			site: 'asos',
			product_id: parse.pathname.replace('/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.bestbuy.com') {
		this.scraper({
			site: 'bestbuy',
			product_id: parse.pathname.replace('/site/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.castorama.fr') {
		this.scraper({
			site: 'castorama',
			product_id: parse.pathname.replace('/store/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.cdiscount.com') {
		this.scraper({
			site: 'cdiscount',
			product_id: parse.pathname + parse.query
		}, callback);
	}
	else if (parse.host == 'www.darty.com') {
		this.scraper({
			site: 'darty',
			product_id: parse.pathname.replace('/','') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.fnac.com') {
		this.scraper({
			site: 'fnac',
			product_id: parse.pathname.replace('/w-4', '').match(/a\d{7}/)
		}, callback);
	}
	else if (parse.host == 'www.grosbill.com') {
		this.scraper({
			site: 'grosbill',
			product_id: parse.pathname.replace('/','') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.laredoute.fr') {
		this.scraper({
			site: 'laredoute',
			product_id: parse.pathname.replace('/ppdp/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.leroymerlin.fr') {
		this.scraper({
			site: 'leroymerlin',
			product_id: parse.pathname.replace('/v3/p/produits/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'livre.fnac.com') {
		this.scraper({
			site: 'livrefnac',
			product_id: parse.pathname.replace('/', '').match(/a\d{7}/)
		}, callback);
	}
	else if (parse.host == 'www.ldlc.com') {
		this.scraper({
			site: 'ldlc',
			product_id: parse.pathname.replace('/fiche/', '') + parse.query
		}, callback);
	}
	else if(parse.host == 'www.rueducommerce.fr')  
	{
		this.scraper({
			site: 'rueducommerce',
			product_id: parse.pathname.replace('/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.target.com') {
		var matches = parse.pathname.match(/\/p\/(.*)/);
		var path = (matches[1]) ? matches[1] : false;

		this.scraper({
			site: 'target',
			product_id: path
		}, callback);
	}
	else if (parse.host == 'www.tesco.com') {

		this.scraper({
			site: 'tesco',
			product_id: parse.pathname.replace('/direct/', '') + parse.query
		}, callback);
	}
	else if (parse.host == 'www.ubaldi.com') {

		this.scraper({
			site: 'ubaldi',
			product_id: parse.pathname.replace('/', '') + parse.query
		}, callback);
	}
	else {
		this.scraper({
			site: false,
			url: url
		}, callback);
	}

}

exports.init = function (url, callback) {
	if (!url || !callback) return false;
	this.parseURL(url, callback);
}
// pass through for parseURL in case architecture changes
